File: phyphox/ExperimentAudioOutput.m
//
//  ExperimentAudioOutput.m
//  phyphox
//
//  Created by Jonas Gessner on 16.03.16.
//  Copyright © 2016 RWTH Aachen. All rights reserved.
//

#import "ExperimentAudioOutput.h"
#import "phyphox-swift.h"
#import <EZAudioiOS/EZAudioiOS.h>

@interface ExperimentAudioOutput () <EZOutputDataSource> {
    EZOutput *_output;
    NSUInteger _lastIndex;
    BOOL _stopPlayback;
    AVAudioPlayer *_player;
    
    BOOL _playing;
}

@end

@implementation ExperimentAudioOutput

- (instancetype)initWithSampleRate:(NSUInteger)sampleRate loop:(BOOL)loop dataSource:(DataBuffer *)dataBuffer {
    self = [super init];
    
    if (self) {
        AudioStreamBasicDescription inputFormat = [EZAudioUtilities monoFloatFormatWithSampleRate:sampleRate];
        _output = [EZOutput outputWithDataSource:self inputFormat:inputFormat];
        
        _dataSource = dataBuffer;
        _sampleRate = sampleRate;
        _loop = loop;
    }
    
    return self;
}

- (OSStatus)output:(EZOutput *)output shouldFillAudioBufferList:(AudioBufferList *)audioBufferList withNumberOfFrames:(UInt32)frames timestamp:(const AudioTimeStamp *)timestamp {
    if (_stopPlayback) {
        _stopPlayback = NO;
        [self pause];
        return -1;
    }
    
    if (_dataSource.count == 0) {
        return noErr;
    }
    
    Float32 *buffer = (Float32 *)audioBufferList->mBuffers[0].mData;
    
    NSArray <NSNumber *> *array = _dataSource.toArray;
    
    for (UInt32 frame = 0; frame < frames; frame++) {
        if (!_loop) {
            if (frame < array.count) {
                buffer[frame] = array[frame].floatValue;
            }
            else {
                _stopPlayback = YES;
                break;
            }
        }
        else {
            NSUInteger index = (_lastIndex+frame) % array.count;
            
            buffer[frame] = [_dataSource objectAtIndexedSubscript:index];
        }
    }
    
    _lastIndex = (_lastIndex+frames-1) % array.count;
    
    return noErr;
}

- (void)play {
    if (!_playing) {
        _playing = YES;
        _stopPlayback = NO;
        [_output startPlayback];
    }
}

- (void)pause {
    if (_playing) {
        _lastIndex = 0;
        [_output stopPlayback];
        _playing = NO;
    }
}

@end

File: phyphox/ExperimentAudioOutput.swift
//
//  ExperimentAudioOutput.swift
//  phyphox
//
//  Created by Jonas Gessner on 22.03.16.
//  Copyright © 2016 RWTH Aachen. All rights reserved.
//

import Foundation

final class ExperimentAudioOutput {
    let dataSource: DataBuffer
    let loop: Bool
    let sampleRate: UInt
    
    var channel: AEBlockChannel!
    
    private var stopPlayback = false
    private var lastIndex = 0
    private var playing = false
    
    init(sampleRate: UInt, loop: Bool, dataSource: DataBuffer) {
        self.dataSource = dataSource;
        self.sampleRate = sampleRate;
        self.loop = loop;
        
        defer {
            channel = AEBlockChannel { [unowned self] (timestamp: UnsafePointer<AudioTimeStamp>, frames: UInt32, data: UnsafeMutablePointer<AudioBufferList>) in
                if dataSource.count == 0 {
                    return
                }
                
                if self.stopPlayback {
                    self.pause()
                    return
                }
                
                let buffer = unsafeBitCast(UnsafeMutableAudioBufferListPointer(data).first!.mData, UnsafeMutablePointer<Float>.self)
                
                let valueArray = dataSource.toArray().map({ (dbl) -> Float in
                    return Float(dbl)
                })
                
                for i in 0..<Int(frames) {
                    if !self.loop {
                        let index = self.lastIndex+i
                        
                        if index < valueArray.count {
                            buffer[i] = valueArray[index]
                        }
                        else {
                            buffer[i] = 0
                            self.stopPlayback = true
                        }
                    }
                    else {
                        let index = (self.lastIndex+i) % valueArray.count
                        
                        buffer[i] = valueArray[index]
                    }
                }
                
                self.lastIndex = (self.lastIndex+Int(frames)-1) % valueArray.count
            }
        }
    }
    
    func play() {
        if !playing {
            playing = true
            stopPlayback = false
            ExperimentManager.sharedInstance().audioController.addChannels([channel])
        }
    }
    
    func pause() {
        if playing {
            lastIndex = 0
            ExperimentManager.sharedInstance().audioController.removeChannels([channel])
            playing = false
        }
    }
    
}
